cranelift/frontend/src/ssa.rs

```diff
@@ -162,6 +170,20 @@ impl SSABuilder {
             && self.results.is_empty()
             && self.side_effects.is_empty()
     }
+
+    fn new_epoch(&mut self) -> Epoch {
+        if self.epoch == Epoch::default() {
+            // The epoch counter wrapped so we need to reset all blocks.
```

Comment:
This code is likely never going to be visited at all, because other implementation limits will prevent us from processing a function of 2^32 blocks. And it'll thus be dead and untested code. I think I'd rather see an assert here that the epoch doesn't wrap (or otherwise do a `checked_add(1).expect("epoch wrapped")` or somesuch).

Reply:
It isn't currently a new epoch per block: it's per call to `use_var` that doesn't find a definition for the variable within the same block, and there could be a lot of those. I'm not sure whether _that_ can happen 2^32 times before hitting other implementation limits. But I'm happy to turn wrapping the counter into a panic, and revisit it later if anyone actually hits it.

Reply:
Ah, I wonder what the impact of using a `u64` instead would be then? That would ensure that for all practical purposes we *never* wrap.

Reply:
I considered it, but there were 7 bytes of padding in `SSABlockData` so a `u32` was the biggest that would avoid making that struct bigger. :laughing:

That said, epoch is only used when the block is sealed, and `undef_variables` can only be non-empty when the block is not yet sealed, so I could overlap them with an enum... :thinking:

And with _that_ said, I should go try a `SecondaryMap<Block, bool>` to store visited flags instead of any of this epoch machinery and see if that easier-to-explain approach actually has any performance impact.

Reply:
Sightglass claims to be able to tell that the epoch counters are faster, by very small margins. According to hyperfine it's either a wash, or when compiling Spidermonkey, the `SecondaryMap` might be very slightly faster. DHAT says the epoch counters allocate slightly less (naturally), read 0.04% fewer bytes, and write 4% fewer bytes (11MiB).

In short, the epoch counters are better, but not enough better to be worth anyone's confusion. I'm switching to a `SecondaryMap` of `bool`.

Reply:
Actually, switching to `EntitySet` now that I discovered that exists. DHAT says this does 0.5% more writes than epoch counters and 0.07% more reads, which is a small enough margin that I don't care. It also executes 0.2% more instructions than `main`, which executes 0.2% more instructions than using epoch counters. 

I've also tested `HashSet` but that executes a lot more instructions as well as doing even more reads and writes.